### PR TITLE
feat(launchpad): add launchpad support devkit and readme

### PR DESCRIPTION
**Pull request by lijunru-hub**
This version includes the following additions:

1. Optional: Added ESP Develop Kits options for each Application. 
    Different development board models such as esp-box and esp-box3 are displayed based on the selected chipset.

2. Optional: Added a description field for each Application.
    This field allows for a brief introduction of the BIN file, including project links and more.

3. Optional: Integrated README.md rendering below each Application.
    Previously, this feature was only displayed after flashing.
    
image

### CODE REVIEW

**Comment by lijunru-hub** [comment]
@RushikeshPatange Hi rushikesh, please try with **flashConfigURL=https://dl.espressif.com/AE/esp-iot-solution/config_test_7_9.toml**

**Comment by lijunru-hub** [comment]
latest effect
image

File: config/config.toml [hunk 1]
```diff
@@ -15,11 +15,18 @@ config_readme_url = "<APP_CONFIG_README_FILE_URL>"
 
 [YOUR_FIRMWARE_APP_NAME]
 # ESP32 Chipsets for which you have a supported Firmware App. Define a property for each supported chipset in the given format.
-chipsets = ["ESP32", "ESP32-S2", "ESP32-C3"]
+chipsets = ["ESP32", "ESP32-S2", "ESP32-C3", "ESP32-S3"]
+# Optional: Set development board for chip model.
+developKits.esp32-s3 = ["esp32-s3-box", "esp32-s3-box3"]
```
Comment (adwait-esp) [inline config/config.toml:20]: We can add in chipsets itself - ESP32-S3-BOX and ESP32-S3-BOX3
`chipsets =  ["ESP32-S3-BOX", "ESP32-S3-BOX3"]`
you can get rid of default example chipsets in your toml file and only keep those that are required.

File: config/config.toml [hunk 2]
```diff
@@ -15,11 +15,18 @@ config_readme_url = "<APP_CONFIG_README_FILE_URL>"
...
 # Configure the name of the binary file
 image.esp32 = "SINGLE_BIN_FOR_ESP32.bin"
 image.esp32-s2 = "SINGLE_BIN_FOR_ESP32-S2.bin"
 image.esp32-c3 = "SINGLE_BIN_FOR_ESP32-C3.bin"
+# Optinal: Configure the name of the image file for development board. Under this configuration, it is not possible to set values for image.esp32-s3 anymore.
+image.esp32-s3.esp32-s3-box = "SINGLE_BIN_FOR_ESP32-S3-BOX.bin"
```
Comment (adwait-esp) [inline config/config.toml:26]: If we define devkits as chipset types then you can directly define.
`image.esp32-s3-box = "SINGLE_BIN_FOR_ESP32-S3-BOX.bin"`
`image.esp32-s3-box3 = "SINGLE_BIN_FOR_ESP32-S3-BOX3.bin"`

File: index.html [hunk 3]
```diff
@@ -200,11 +200,40 @@ <h5 class="pt-4 pb-2 lh-lg"><span id="qtLabel">Choose from some of ESP's pre-bui
...
                         </div>
+                        <div class="field-container" id ="developKitsContainer" style="display: none">
+                            <label>
+                                ESP Develop Kits
+                            </label>
+                            <div class="field">
+                                <div id="developKits">
+                                    <div class="form-check-inline">
+                                        <label class="form-check-label" for="radio3">ESP32-S3-DevKits&nbsp;
+                                            <input type="radio" checked class="form-check-input" name="developKitsType" id="radio3" value="esp32_matter_all-clusters-app_te6_merged.bin">
```
Comment (adwait-esp) [inline index.html:211]: Why are these added here matter_all_clusters-app?

**Review by adwait-esp — COMMENTED** [review]
Few comments. rest changes for readme, app description looks good

File: config/config.toml (hunk 1 above)
Comment (lijunru-hub) [reply config/config.toml:20]: This approach is indeed perfect, as it eliminates a lot of complex logic. We could also add a global definition to modify the title "ESP Chipset Type." This way, it can indicate whether it's for chip or development board selection.

File: index.html (hunk 3 above)
Comment (lijunru-hub) [reply index.html:211]: I forgot to make the changes.

File: config/config.toml (hunk 1 above)
Comment (lijunru-hub) [reply config/config.toml:20]: After discussion, it has been modified as follows:

```
chipsets = ["ESP32", "ESP32-S2", "ESP32-C3", "ESP32-S3"]
# Optional: Set development board for chip model.
developKits.esp32-s3 = ["esp32-s3-box", "esp32-s3-box3"]
# Configure the name of the binary file.
image.esp32 = "SINGLE_BIN_FOR_ESP32.bin"
image.esp32-s2 = "SINGLE_BIN_FOR_ESP32-S2.bin"
image.esp32-c3 = "SINGLE_BIN_FOR_ESP32-C3.bin"
# Optional: Configure the name of the image file for the development board. Under this configuration, image.esp32-s3 will become invalid.
image.esp32-s3-box = "SINGLE_BIN_FOR_ESP32-S3-BOX.bin"
image.esp32-s3-box3 = "SINGLE_BIN_FOR_ESP32-S3-BOX3.bin"
```

File: index.html (hunk 3 above)
Comment (lijunru-hub) [reply index.html:211]: Modify it to not assign values by default.

**Comment by lijunru-hub** [comment]
@adwait-esp Please try with laster https://dl.espressif.com/AE/esp-iot-solution/config_test_7_16.toml

**Review by adwait-esp — APPROVED** [review]

**Comment by RushikeshPatange** [comment]
@lijunru-hub, thanks for your contribution!

I have tested the feature with the latest [TOML URL](https://dl.espressif.com/AE/esp-iot-solution/config_test_7_16.toml) you mentioned. Here are some suggestions for improvement:

1. **UI Distortion Issue**: We have an existing feature that shows application configuration, which can be seen [here](https://github.com/espressif/esp-launchpad/pull/62). However, the implementation of the `appDescription` container is causing UI distortion, as shown below:
   feature_implementation

2. **Flash Progress Message**: The flash progress message container gets pushed way down, as most of the height is taken by the `appDescription` container. This requires users to scroll a lot, which can be inconvenient for the UX.

3. **QR Code Display**: We have the ability to show QR codes at the bottom of the **Quick Start** page, which are used by iOS and Android apps. These should be displayed if any `android_app_url` or `ios_app_url` is provided in the TOML under `supported_app`.

4. **Flash Button Enablement**: If the chipset has devkits, can we enable the flash button only after a devkit selection for that chip? Currently, it's giving a wrong error message, which can be avoided by enabling the flash button only after devkit selection. In a normal scenario, the flash button will be enabled after chipset selection.
   image

---

### Approach Suggestion (for points 1 to 3)

**Before Show Application Information Click:**
Screenshot 2024-07-18 at 14 18 44

**After Show Application Information Click:**
Screenshot 2024-07-18 at 14 19 26

I have used the Bootstrap offcanvas component and wrapped the appDescription container inside it which we can toggle with `Show Application Information` link, as we are using Bootstrap in our launchpad. I believe this approach can achieve the following:

- Avoid the UI distortion mentioned in point 1.
- Ensure users clearly see the flash progress status message, with the option to view app info if needed.
- Keep the QR code containers within the user's viewport.
- Reduce the amount of scrolling required, enhancing the overall UX.

Could you please review and fix point 4 and the minor comments added? 
I will proceed with the implementation of the mentioned approach after merging this MR, including point 4 and the suggested fixes

File: js/index.js [hunk 4]
```diff
@@ -191,8 +227,49 @@ function populateDeviceTypes(imageConfig) {
     });
 }*/
 
+function populateSupportedDevelopKits(developKitsConfig) {
+    developKitsRadioGroup.innerHTML = "";
+    let i = 1;
```
Comment (RushikeshPatange) [inline js/index.js:232]: Could we avoid using the variable i since it is not used anywhere in the function?

**Review by RushikeshPatange — CHANGES_REQUESTED** [review]

**Comment by lijunru-hub** [comment]
> @lijunru-hub, thanks for your contribution!
> 
> I have tested the feature with the latest [TOML URL](https://dl.espressif.com/AE/esp-iot-solution/config_test_7_16.toml) you mentioned. Here are some suggestions for improvement:
> 
> 1. **UI Distortion Issue**: We have an existing feature that shows application configuration, which can be seen [here](https://github.com/espressif/esp-launchpad/pull/62). However, the implementation of the `appDescription` container is causing UI distortion, as shown below:
>    feature_implementation
> 2. **Flash Progress Message**: The flash progress message container gets pushed way down, as most of the height is taken by the `appDescription` container. This requires users to scroll a lot, which can be inconvenient for the UX.
> 3. **QR Code Display**: We have the ability to show QR codes at the bottom of the **Quick Start** page, which are used by iOS and Android apps. These should be displayed if any `android_app_url` or `ios_app_url` is provided in the TOML under `supported_app`.
> 4. **Flash Button Enablement**: If the chipset has devkits, can we enable the flash button only after a devkit selection for that chip? Currently, it's giving a wrong error message, which can be avoided by enabling the flash button only after devkit selection. In a normal scenario, the flash button will be enabled after chipset selection.
>    image
> 
> ### Approach Suggestion (for points 1 to 3)
> **Before Show Application Information Click:** Screenshot 2024-07-18 at 14 18 44
> 
> **After Show Application Information Click:** Screenshot 2024-07-18 at 14 19 26
> 
> I have used the Bootstrap offcanvas component and wrapped the appDescription container inside it which we can toggle with `Show Application Information` link, as we are using Bootstrap in our launchpad. I believe this approach can achieve the following:
> 
> * Avoid the UI distortion mentioned in point 1.
> * Ensure users clearly see the flash progress status message, with the option to view app info if needed.
> * Keep the QR code containers within the user's viewport.
> * Reduce the amount of scrolling required, enhancing the overall UX.
> 
> Could you please review and fix point 4 and the minor comments added? I will proceed with the implementation of the mentioned approach after merging this MR, including point 4 and the suggested fixes

Hi @RushikeshPatange 

I have added the following check for the fourth point: if a developKits selection exists, you must choose both the chipset and developKits options before proceeding with the flash.

```c
    if(chipSetsRadioGroup.querySelectorAll("input[type=radio]:checked").length!== 0 &&
       (developKitsContainer.style.display === "none" ||
       developKitsRadioGroup.querySelectorAll("input[type=radio]:checked").length!== 0)){
```

I think the suggestions for points 1 to 3 are excellent. Thank you!

**Comment by RushikeshPatange** [comment]
Hi @lijunru-hub ,

You have added checks, that's fine, but lets consider this scenario, if user selected the chipset (which has devkits) and directly jump on flash button click then your check logic will make the code to fall in else part which show **alertChipsetSelectMsg**, showing wrong error message (as you can see in the attachment of point 4) of **Unable to flash device. Please ensure that chipset type is selected before flashing.** even user has selected the chipset, so in context of UX it would be better to enable `flash button` after both chipset and devkit selection (as previously mentioned in 4th point) that avoid **user extra click**.

Also suggested one change to avoid using the variable `i` since it is not used anywhere in the `populateSupportedDevelopKits` function

Could you please review these suggestions and implement the necessary changes?

**Comment by RushikeshPatange** [comment]
@lijunru-hub, LGTM with latest changes. Thanks for the contribution 😄
